catalog/app/utils/json-schema/json-schema.ts

```diff
@@ -0,0 +1,221 @@
+import Ajv from 'ajv'
+import * as dateFns from 'date-fns'
+import * as R from 'ramda'
+
+type CompoundCondition = 'anyOf' | 'oneOf' | 'not' | 'allOf'
+
+export type JsonSchema = Partial<
+  {
+    $ref: string
+    const: string
+    dateformat: string
+    default: any
+    enum: $TSFixMe[]
+    format: string
+    items: JsonSchema
+    properties: Record<string, JsonSchema>
+    type: string | string[] | JsonSchema[]
+  } & Record<CompoundCondition, JsonSchema[]>
+>
+
+// NOTE: can't use R.prop directly, because R.prop type doesn't allow undefined
+const prop = (
+  property: keyof JsonSchema,
+  optSchema?: JsonSchema,
+): JsonSchema[keyof JsonSchema] => (optSchema ? R.prop(property, optSchema) : undefined)
+
+export const isSchemaArray = (optSchema?: JsonSchema) =>
+  prop('type', optSchema) === 'array'
+
+export const isSchemaObject = (optSchema?: JsonSchema) =>
+  prop('type', optSchema) === 'object'
+
+const isSchemaString = (optSchema?: JsonSchema) => prop('type', optSchema) === 'string'
+
+const isSchemaNumber = (optSchema?: JsonSchema) => prop('type', optSchema) === 'number'
+
+const isSchemaInteger = (optSchema?: JsonSchema) => prop('type', optSchema) === 'integer'
+
+export const isSchemaBoolean = (optSchema?: JsonSchema) =>
+  prop('type', optSchema) === 'boolean'
+
+const isSchemaNull = (optSchema?: JsonSchema) => prop('type', optSchema) === 'null'
+
+export const isSchemaEnum = (optSchema?: JsonSchema) => !!prop('enum', optSchema)
+
+export const isSchemaOneOf = (optSchema?: JsonSchema) => !!prop('oneOf', optSchema)
+
+export const isSchemaAnyOf = (optSchema?: JsonSchema) => !!prop('anyOf', optSchema)
+
+export const isSchemaAllOf = (optSchema?: JsonSchema) => !!prop('allOf', optSchema)
+
+const isSchemaConst = (optSchema?: JsonSchema) => !!prop('const', optSchema)
+
+function isSchemaCompound(optSchema?: JsonSchema) {
+  if (!optSchema) return false
+  return ['anyOf', 'oneOf', 'not', 'allOf'].some(
+    (key) => optSchema[key as 'anyOf' | 'oneOf' | 'not' | 'allOf'],
+  )
+}
+
+const isSchemaReference = (optSchema: JsonSchema) => !!prop('$ref', optSchema)
+
+export const isNestedType = R.either(isSchemaArray, isSchemaObject)
+
+function compoundTypeToHumanString(
+  optSchema: JsonSchema,
+  condition: CompoundCondition,
+  divider: string,
+): string {
+  if (!Array.isArray(prop(condition, optSchema))) return ''
+
+  return (optSchema[condition] as JsonSchema[])
```

Comment:
why is it necessary to cast the value to `JsonSchema[]` here? seems like it should be resolved to that automatically. if that's because it's possibly undefined, we could just use a trailing `!`